From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sun, 30 Jul 2023 12:37:28 +0200
Subject: [PATCH] Client perspective - Deep replacements - Method to send
 updates

Adds a method to send the necessary updates to a player when
the way they observe deep replaceables may have changed.

License: Fiddle Public License Agreement 1.0 (included in license/FPL-1.0.txt)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index f687bb6b8899885f918c309edc4b1fd4c1dca5c9..fea2a32d452133bea967f3ee1eb4bce385e80527 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2677,4 +2677,32 @@ public class ServerPlayer extends Player {
     }
     // Fiddle end - client perspective - player
 
+    // Fiddle start - client perspective - deep replacements - send updates
+    /**
+     * Sends all necessary packets to this player to update any deep replaceable content
+     * after the content that the client should observe may have changed.
+     * <br>
+     * This method should be called whenever something about the player or their client changes that influences
+     * deep replacements.
+     */
+    public void sendUpdatesForDeepReplaceableContent(boolean itemsMayHaveChanged, boolean translationsMayHaveChanged) {
+
+        // Items contain items, so we make sure that this implication is set correctly
+        itemsMayHaveChanged |= translationsMayHaveChanged;
+
+        if (itemsMayHaveChanged) {
+            // Resend the player's own inventory contents
+            this.getBukkitEntity().updateInventory();
+        }
+
+        // TODO if itemsMayHaveChanged, re-send tracked entity synched data
+        // TODO if itemsMayHaveChanged, re-send all visible block entity data
+        // TODO if itemsMayHaveChanged, re-send contents of other inventory that player has open
+        // TODO if translationsMayHaveChanged, update title of inventory that player has open
+        // TODO if translationsMayHaveChanged, update other visible texts to player like boss bar text
+        // TODO if translationsMayHaveChanged, update book content if they have one open maybe?
+
+    }
+    // Fiddle end - client perspective - deep replacements - send updates
+
 }
